Thinking budget for Google Ai studio (Gemini 2.5 flash)

LeZeez opened the pull request:
<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

Second attempt, hopefully this works fine.

This adds a `Enable thinking` and `Set thinking budget` checkboxes for Gemini 2.5 flash (And future models) with a slider for `Thinking Budget (tokens)`

This behaves exactly like Google AI studio:

`Enable thinking` is ON, `Set thinking budget` is OFF = Default thinking without a budget.
`Enable thinking` is ON, `Set thinking budget` is ON  = Uses the `Thinking Budget (tokens)` slider value.
`Enable thinking` is OFF = No thinking. (thinking budget = 0)

This only appears/works for Gemini 2.5 flash.

LeZeez committed:
Old commits.

Old, left for context.

LeZeez commented:
I just realized that VS code auto formatting added all these lines in the index.html I hope this isn't a big problem

bmen25124 commented:
Since commits are a little messy, we should squash & merge them. @Cohee1207

Cohee1207 commented:
1. What's the point of `masterEnableThinking`? There is already a `show_thoughts` toggle that is hidden for Gemini
2. Why not use an existing `reasoning_effort` setting? It's not a slider but it's how other sources are doing it. Pre-allocate same sane values for 3 available settings (high = max, mid = half, low = minimum).

Cohee1207 commented:
Proposals (for effort to budget mapping):
1. high = 24000
2. mid = 10000
3. low = 5000

lyncmdos commented:
Maybe add a "close" option? I really want the thinking mode closed. And I think Slider may be better beacuse it's just like use Gemini in Google AI Studio, users are more customized.

lyncmdos commented:
> Proposals (for effort to budget mapping):
> 1. high = 24000
> 2. mid = 10000
> 3. low = 5000

Maybe add a "close" option? I really want the thinking mode closed. And I think Slider may be better beacuse it's just like use Gemini in Google AI Studio, users are more customized, and it's requested in [issues 3875](https://github.com/SillyTavern/SillyTavern/issues/3875), shows it's liked for more people.

LeZeez commented:
> What's the point of masterEnableThinking? There is already a show_thoughts toggle that is hidden for Gemini

Thanks for the feedback.

 `masterEnableThinking` is to enable/disable the **thinking budget** param only and is not related to `show_thoughts`

>Why not use an existing reasoning_effort setting? It's not a slider but it's how other sources are doing it. Pre-allocate same sane values for 3 available settings (high = max, mid = half, low = minimum).

As for the `reasoning_effort` I believe it's better to make it slider like the AI studio, and it's requested [here](https://github.com/SillyTavern/SillyTavern/issues/3875)

@Cohee1207

LeZeez commented:
> Maybe add a "close" option? I really want the thinking mode closed.

There is already a checkbox `Enable thinking budget` that controls the param being added or not (so you can use the default thinking if it's unchecked), and if it's checked there is another checkbox `Use thinking budget value` to use the slider or not (If not, thinking budget = 0, which is no thinking, and If it's used, then the slider value is being used)
Screenshot:
![image](https://github.com/user-attachments/assets/65fdfbd6-6484-4da5-b656-200cc05078d3)

@lyncmdos

lyncmdos commented:
> > Maybe add a "close" option? I really want the thinking mode closed.
> 
> There is already a checkbox `Enable thinking budget` that controls the param being added or not (so you can use the default thinking if it's unchecked), and if it's checked there is another checkbox `Use thinking budget value` to use the slider or not (If not, thinking budget = 0, which is no thinking, and If it's used, then the slider value is being used)
> Screenshot:
> ![image](https://github.com/user-attachments/assets/65fdfbd6-6484-4da5-b656-200cc05078d3)
> 
> @lyncmdos 

Thanks a lot, I've git switched to this commit, and have used it for half an hour, I like it very much. I really need the function, just can't wait for a minute.

equal-l2 commented:
EDIT: Oops, I didn't read the post just above, disregard please...

After some experiments with the API, I found that thinking is enabled for Gemini 2.5 Flash by default, with automatic budget setting or some default value.
To turn off thinking, you must set "thinking_budget" to 0.

The reason I think that way: it takes 2.5~3s to get a response without setting `thinking_budget`, while it takes 1.5s with `thinking_budget=0`.

equal-l2 commented:
On second thought, there might be a few helpful points in my previous post.
1. Thinking is enabled by default, and **you must explicitly turn it off if you don't want.**
2. If you are fine with the default, you don't need to set thinking_budget. (the current code sets `1000` as own default)

cloak1505 commented:
[retracted]

Edit: If API is different from AI Studio, my bad. I typed up above in a panic without having tested or touched the branch yet.

Edit 2: I see. Disregard my entire comment, assuming API enables thinking by default. Not sure how I feel about this fact but it works.

LeZeez commented:
> After some experiments with the API, I found that thinking is enabled for Gemini 2.5 Flash by default, with automatic budget setting or some default value. To turn off thinking, you must set "thinking_budget" to 0.
> 
> The reason I think that way: it takes 2.5~3s to get a response without setting `thinking_budget`, while it takes 1.5s with `thinking_budget=0`.

Thanks for the feedback. I already mentioned that and provided a screenshot [here](https://github.com/SillyTavern/SillyTavern/pull/3880#issuecomment-2816739116)
And by the way, if you want to test the thinking is working or not you can set the max-output to something little like 50-100 then if you get cut response that is mostly from the thinking process, then it's working. If you get an actual response that is lower than the max-output or a cut response, then thinking isn't working. 

### Examples:
Thinking on (Default, `Enable thinking budget` disabled, max-output = 80)

![image](https://github.com/user-attachments/assets/ce33ff47-961b-47c4-994d-60d7f415bbcd)

Thinking on (`Enable thinking budget` and `Use thinking budget value` enabled, Thinking Budget = 1k, max-output = 80)

![image](https://github.com/user-attachments/assets/7c579e72-011f-4316-b235-9c48195dbcc6)

Thinking off (`Enable thinking budget` is enabled, `Use thinking budget value` disabled, meaning Thinking Budget = 0, max-output = 80)

![image](https://github.com/user-attachments/assets/9c12e03f-93f2-433c-965d-6705582e7326)

You can also set the slider to `0` !

Edit: I'll leave this for context too

LeZeez commented:
~~>TL;DR: Remove "Use thinking budget value" checkbox. *[deleted]*~~

~~I believe it's better like this so if you already set a value you can just disable thinking then reuse it again instead of writing the value then `0` then the value again.~~

~~>Also, why are we capping thinking budget to 24000 instead of 24576? *[deleted]*~~

~~The `Unlocked Context Size` unlocks it and makes it adjustable up to `100k` for future updates.~~


~~Edit: sorry for late replies. I'll leave this comment for context.~~

cloak1505 commented:
> Edit 2: I see. Disregard my entire comment, assuming API enables thinking by default. Not sure how I feel about this fact but it works.

Well OK but this PR's options as-is is oddly inverted name-wise when you ignore the descriptions about disablement.

I stand by my original opinion that the first checkbox should be "Enable thinking mode". This doesn't require any description aside from the little pop-up saying it only works on 2.5 Flash Preview.

When "Enable thinking mode" is disabled, set `thinkingBudget` to 0 and hide the checkbox and slider below. Intuitive.

Second checkbox: "Set thinking budget" with description `When disabled, let the model decide how many thinking tokens to use. 0 disables thinking mode.`

There are exactly four use cases, in order:
```
Thinking mode is OFF, thinkingBudget is 0 (does not change slider)
[ ] Enable thinking mode

Thinking mode is ON, thinkingBudget is NOT sent (let model decide)
[x] Enable thinking mode
[ ] Set thinking budget
Thinking Budget: 2000

Thinking mode is ON, thinkingBudget is >=1 from slider
[x] Enable thinking mode
[x] Set thinking budget
Thinking Budget: 2000

Thinking mode is OFF, thinkingBudget is 0 from slider
[x] Enable thinking mode
[x] Set thinking budget
Thinking Budget: 0
```

Compare with the current PR:
```
Thinking mode is ON, thinkingBudget is NOT sent (let model decide)
[ ] Enable thinking budget

Thinking mode is OFF, thinkingBudget is 0 (does not change slider)
[x] Enable thinking budget
[ ] Use thinking budget value
Thinking Budget: 2000

Thinking mode is ON, thinkingBudget is >=1 from slider
[x] Enable thinking budget
[x] Use thinking budget value
Thinking Budget: 2000

Thinking mode is OFF, thinkingBudget is 0 from slider
[x] Enable thinking budget
[x] Use thinking budget value
Thinking Budget: 0
```

Notice how you should mostly know what the setup on top means without reading descriptions. The order of bottom setup is slightly different - thinking is turned off by turning off "Use thinking budget value" but the wording "Enable thinking budget" is on.

---

One last thing, lowercase 'tokens' in "Thinking Budget (tokens)" for consistency with Context Size and Max Response Length sliders.

LeZeez commented:
> > Edit 2: I see. Disregard my entire comment, assuming API enables thinking by default. Not sure how I feel about this fact but it works.
> 
> Well OK but this PR's options as-is is oddly inverted name-wise when you ignore the descriptions about disablement.
> 
> I stand by my original opinion that the first checkbox should be "Enable thinking mode". This doesn't require any description aside from the little pop-up saying it only works on 2.5 Flash Preview.
> 
> When "Enable thinking mode" is disabled, set `thinkingBudget` to 0 and hide the checkbox and slider below. Intuitive.
> 
> Second checkbox: "Set thinking budget" with description `When disabled, let the model decide how many thinking tokens to use. 0 disables thinking mode.`
> 
> There are exactly four use cases, in order:
> 
> ```
> Thinking mode is OFF, thinkingBudget is 0 (does not change slider)
> [ ] Enable thinking mode
> 
> Thinking mode is ON, thinkingBudget is NOT sent (let model decide)
> [x] Enable thinking mode
> [ ] Set thinking budget
> Thinking Budget: 2000
> 
> Thinking mode is ON, thinkingBudget is >=1 from slider
> [x] Enable thinking mode
> [x] Set thinking budget
> Thinking Budget: 2000
> 
> Thinking mode is OFF, thinkingBudget is 0 from slider
> [x] Enable thinking mode
> [x] Set thinking budget
> Thinking Budget: 0
> ```
> 
> Compare with the current PR:
> 
> ```
> Thinking mode is ON, thinkingBudget is NOT sent (let model decide)
> [ ] Enable thinking budget
> 
> Thinking mode is OFF, thinkingBudget is 0 (does not change slider)
> [x] Enable thinking budget
> [ ] Use thinking budget value
> Thinking Budget: 2000
> 
> Thinking mode is ON, thinkingBudget is >=1 from slider
> [x] Enable thinking budget
> [x] Use thinking budget value
> Thinking Budget: 2000
> 
> Thinking mode is OFF, thinkingBudget is 0 from slider
> [x] Enable thinking budget
> [x] Use thinking budget value
> Thinking Budget: 0
> ```
> 
> Notice how you should mostly know what the setup on top means without reading descriptions. The order of bottom setup is slightly different - thinking is turned off by turning off "Use thinking budget value" (it _is_ using value of 0) but the wording "Enable thinking budget" is on.
> 
> One last thing, lowercase 'tokens' in "Thinking Budget (tokens)" for consistency with Context Size and Max Response Length sliders.

Thanks for your feedback. Regarding the naming this can be changed easily, But the other things would require some time again.
I'll tell you why this is a good approach (at least for me):
- This is a new param and it only works with 2.5 flash, so your suggestion for thinking wouldn't work on other models (Like 2.0 thinking and 2.5 pro). This can be edited with future models, but right now it's good like this.
- I made the master checkbox to enable/disable the param itself, meaning that if it's not checked then it'll use the default thinking mode (Like AI studio but inverted with 2.5 flash, but exactly like 2.0 thinking & 2.5 pro as it uses the default or automatic one)
- The second check box one is good (maybe the naming isn't) as it separates using thinking or not (This one is the similar to AI studio, but also inverted like what you said.) So if it's checked you can use the slider. Explained [here](https://github.com/SillyTavern/SillyTavern/pull/3880#issuecomment-2816759981) too.
- Your approach would be excellent if all the models support this.

Tell me what do you think about it and we can change anything if needed. Names will change when finding good suggestions or waiting others to review it too.

cloak1505 commented:
@LeZeez I now have a better picture of your intent.

Revised descriptions for clarity of the first option:

```
[ ] Enable thinking budget
Sends a `thinkingBudget` value. If unchecked, lets the model decide (default thinking mode).

[ ] Use thinking budget value
Uses the slider value below. If unchecked, the budget is explicitly set to 0 (thinking disabled).
```

Edit: Using "Use system prompt" as an example, the span code would look like this:

```
<span data-i18n="thinking_budget_desc_1">
Sends a</span> <code>thinkingBudget</code> <span data-i18n="thinking_budget_desc_2">value. If unchecked, lets the model decide (default thinking mode).</span>
```

LeZeez commented:
![image](https://github.com/user-attachments/assets/58849b13-94e8-4e4c-bed0-af9677fc73ab)

All good?

cloak1505 commented:
Just lowercase tokens, but yeah all looks good to me.

Edit: Half the real answer to my early question about the budget cap is that without "Unlocked Context Size" or "Mad Lab Mode", the slider increments by 100, which is good. In this case, 24500 is technically closer to the real cap.

LeZeez commented:
> Just lowercase tokens, but yeah all looks good to me.
> 
> Edit: Half the real answer to my early question about the budget cap is that without "Unlocked Context Size" or "Mad Lab Mode", the slider increments by 100, which is good. In this case, 24500 is technically closer to the real cap.

Check the new commit and give me your feedback. @cloak1505

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -439,7 +447,11 @@ async function sendMakerSuiteRequest(request, response) {
         });
 
-        const apiVersion = isThinking ? 'v1alpha' : 'v1beta';
+        let apiVersion = isThinking ? 'v1alpha' : 'v1beta';
```
`isThinking || generationConfig.thinkingConfig ? 'v1alpha' : 'v1beta'`

cloak1505 replied:
What's v1alpha for? No other place in ST uses v1alpha, and "thinkingConfig" works when I sent a manual v1beta request. Edit: I saw #commit-log in Discord; he took the advice from an LLM, and another user noted the [doc](https://ai.google.dev/gemini-api/docs/thinking#set-budget) uses v1beta in the example.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -354,6 +354,14 @@ async function sendMakerSuiteRequest(request, response) {
+        generationConfig.thinkingConfig = {
+            thinkingBudget: thinkingConfigFromRequest.thinkingBudget,
+        };
+        console.log('[chat-completions.js] Added thinkingConfig to generationConfig:', generationConfig.thinkingConfig);
```
Useless log.

LeZeez replied:
Oh! Forgot to remove this. 2.5 pro put a lot of logs. mb.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -387,6 +390,9 @@ const default_settings = {
     request_images: false,
     seed: -1,
     n: 1,
+    masterEnableThinking: false,
```
New values don't follow an existing naming pattern. Also, I really find its naming confusing.

cloak1505 replied:
Could be `send_thinking_budget` instead since it literally means we're sending a value whether 0 or positive. ~~For accuracy, the second one could be `positive_thinking_budget` since off is really 0 as opposed to nothing at all.~~ Whoops, inaccurate since you can also set slider to 0. just `use_thinking_budget` then. Sounds suspiciously similar to the first, but it be like that.

cloak1505 replied:
My earlier suggested setup would be `enable_thinking` and `set_thinking_budget` which are clear on what they do.

Cohee1207 replied:
`use_thinking_budget` with `thinking_budget` and no second checkbox would be closer to ideal.

cloak1505 commented:
Does not error on 2.5-pro or 2.0-flash-thinking but response time does not shorten for 0 budget. But we need to exclude `thinkingConfig` from older non-thinking models.

2.0-flash, 1.5-pro
```
Google AI Studio API returned error: 400 Bad Request {
  "error": {
    "code": 400,
    "message": "Unable to submit request because thinking_budget is only supported when enable_thinking is true. Learn more: https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/gemini",
    "status": "INVALID_ARGUMENT"
  }
}
```

2.0-flash-lite
```
Google AI Studio API returned error: 400 Bad Request {
  "error": {
    "code": 400,
    "message": "thinking_config is not supported.",
    "status": "INVALID_ARGUMENT"
  }
}
```

1.5-flash-8b
```
Google AI Studio API returned error: 400 Bad Request {
  "error": {
    "code": 400,
    "message": "Thinking is not enabled for models/gemini-1.5-flash-8b-exp-0827",
    "status": "INVALID_ARGUMENT"
  }
}
```

Cohee1207 commented:
Why a second checkbox? The common use pattern when you want to disable a slider setting is to set it to default value. If you want to set a 0 value to a slider - set it. Same as penalties, temperature, etc. They all have "neutral" or "disabled" values.

LeZeez commented:
>Does not error on 2.5-pro or 2.0-flash-thinking but response time does not shorten for 0 budget. But we need to exclude thinkingConfig from older non-thinking models.

yeah I was just about to tell you about this.

Cohee1207 commented:
> But we need to exclude thinkingConfig from older non-thinking models.

Great, another `if (modelName)`. I love it.

LeZeez commented:
> Why a second checkbox? The common use pattern when you want to disable a slider setting is to set it to default value. If you want to set a 0 value to a slider - set it. Same as penalties, temperature, etc. They all have "neutral" or "disabled" values.

I thought it would be better to make things somewhat easier, but I doubt the easy part tbh. Instead of switching to 0 then a value, you can just use that checkbox so you don't have to write 0 manually every time.

cloak1505 commented:
> Why a second checkbox? The common use pattern when you want to disable a slider setting is to set it to default value. If you want to set a 0 value to a slider - set it. Same as penalties, temperature, etc. They all have "neutral" or "disabled" values.

Because the first checkbox doesn't turn thinking off but rather determines if `thinkingBudget` is sent (if not sent, model still thinks, and decides on its own), and he's using second checkbox off as actually-off. I originally argued for the first checkbox to be "Enable thinking" so off is off (0 budget).

> so you don't have to write 0 manually every time.

We wouldn't have to if we went the "Enable thinking" route.

LeZeez commented:
>We wouldn't have to if we went the "Enable thinking" route.

Fair enough. Maybe that will change when all models will be thinking (maybe sooner).

cloak1505 commented:
> Fair enough. Maybe that will change when all models will be thinking (maybe sooner).

I don't believe how many models are able to toggle thinking is too relevant. Similarly, not all models take a thinking budget.

If we really wanted to be super clear, it can look like this:

```
[ ] Enable thinking
(Gemini 2.5 Flash Preview is the only thinking model that can have its thinking disabled.)
```
so it's not tucked behind the hover tooltip. As mentioned before, the meaning of the option is clear enough that it doesn't need an explanation on what it does.

Instead of changing options in the future, it may be better to set it up this way from the start.

Edit: Side effect is there would still be two checkboxes, not saving more space than Cohee's preference to remove the second checkbox (maybe that's more to avoid confusing names).

LeZeez commented:
I'm really sorry I won't be able to continue this PR mostly, I hope you understand as I'm real busy these days and this PR was just an attempt to do something. If it's not good then someone else that has better experience than me can do it🥲

Cohee1207 commented:
No need to rush with it, it needs to be done right. No one is dying without this merged ASAP. If they do - they could pull the PR locally.

LeZeez commented:
Just finished the one with Enable thinking and proper naming now. testing...

LeZeez committed:
Fresh start.

Fresh start. Added controls that are similar to Google AI studio. Only visible for 2.5 flash.

LeZeez commented:
Oh forgot about this🥲

Cohee1207 commented on public/index.html:
```diff
@@ -2034,6 +2034,45 @@ <h4 class="range-block-title justifyCenter">
                                             </span>
                                         </div>
                                     </div>
+                                    <div id="thinking_budget_master_container" data-source="makersuite" style="display: none;">
```
Why the display: none? It's set automatically if the source is not makersuite.

LeZeez replied:
Sorry for that too. will clean it up.

Cohee1207 commented on public/index.html:
```diff
@@ -2034,6 +2034,45 @@ <h4 class="range-block-title justifyCenter">
+                                                        <input class="neo-range-input" type="number" min="0" max="24576" step="1" data-for="thinking_budget_slider" id="thinking_budget_counter" value="1000">
+                                                    </div>
+                                                </div>
+                                                <label for="oai_max_context_unlocked" class="checkbox_label displayNone"></label>
```
What's the point of this label? Also I don't see any reason to bind unlocked _context size_ with it.

LeZeez replied:
To be honest sometimes I just follow 2.5 pro blindly so don't mind me. Will clean it up.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -165,6 +165,11 @@ const textCompletionModels = [
     'code-search-ada-code-001',
 ];
 
+const googleThinkingBudgetModels = [
```
Model lists... So many model lists...

<img src="https://github.com/user-attachments/assets/492419f5-1a83-47f8-a847-855a3a591a3a" width=200 />

cloak1505 replied:
Cohee, what do you think of the idea of generalizing `enable_thinking` and `set_thinking_budget`? Then this list can be `thinkingBudgetModels` in case it can be shared with other APIs in the future.

Cohee1207 replied:
> Cohee, what do you think of the idea of generalizing enable_thinking and set_thinking_budget?

Generalizing as removing `set_thinking_budget`? Sure, I'm all in for it.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -301,6 +306,9 @@ export const settingsToUpdate = {
     show_thoughts: ['#openai_show_thoughts', 'show_thoughts', true],
     reasoning_effort: ['#openai_reasoning_effort', 'reasoning_effort', false],
+    google_master_enable_thinking: ['#google_master_enable_thinking', 'google_master_enable_thinking', true],
+    google_set_thinking_budget: ['#google_set_thinking_budget', 'google_set_thinking_budget', true],
```
So we still going with two checkboxes?

LeZeez replied:
Wanted to make it exactly like AI studio...

cloak1505 replied:
That was my original intent too. Hypothetically we can roll the second checkbox and thinking budget as a single Reasoning Effort dropdown with an Auto option but then this will collide with other APIs not supporting Auto e.g. 3.7 Sonnet requires a budget.

cloak1505 replied:
Whoa, would it be possible to add a checkbox to Reasoning Effort for Gemini only? Internally it will be tied to set_thinking_budget to decide whether the % setting is used.

```
[ ] Reasoning Effort (tooltip: blah blah but maximum 24576 tokens.)  [High ▼]
If unchecked, the budget is automatically determined by the API.
```

Cohee1207 replied:
I don't want a second checkbox. It's redundant. AI Studio can do everything they want because all they ever have to care about is their own models. We have 20 backends to care about.

LeZeez replied:
I know it may seem redundant, but it's the only way to also make requests without `thinkingBudget` that it'll use the default thinking. If it's a checkbox and a slider only, we can't do that. What do you think?

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2092,8 +2106,20 @@ async function sendOpenAIRequest(type, messages, signal) {
-        generate_data['stop'] = getCustomStoppingStrings(stopStringsLimit).slice(0, stopStringsLimit).filter(x => x.length >= 1 && x.length <= 16);
+        const generationConfig = {};
+        generationConfig['top_k'] = Number(oai_settings.top_k_openai);
+        generationConfig['stopSequences'] = getCustomStoppingStrings(stopStringsLimit).slice(0, stopStringsLimit).filter(x => x.length >= 1 && x.length <= 16);
```
I think you just broke stop strings.

LeZeez replied:
Yeah 2.5 pro actually edited that and I noticed it after the commit so I just waited for review

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2092,8 +2106,20 @@ async function sendOpenAIRequest(type, messages, signal) {
+        if (googleThinkingBudgetModels.includes(model)) {
+            if (!oai_settings.google_master_enable_thinking) {
+                // Master switch is OFF: Explicitly disable thinking
+                generationConfig['thinkingConfig'] = { thinkingBudget: 0 };
```
Why sending it as an object? Maybe just a single number? Since there's nothing else in there.

cloak1505 replied:
> Why sending it as an object?

What's the question here?

![python request](https://github.com/user-attachments/assets/fa721994-918c-4bbc-8a31-bae84ec6f044)

Cohee1207 replied:
The question is the file that I'm commenting on is from a frontend. You don't have to follow backend's conventions here.

cloak1505 replied:
I don't know anything about js/ts, how else would this be written if thinkingBudget is required to be inside thinkingConfig inside generationConfig?

Edit: `generationConfig['thinkingConfig']['thinkingBudget'] = 0;` but doesn't this require the thinkingBudget to exist first?

Cohee1207 replied:
You don't have to put thinkingConfig into generationConfig as an object, it's an internal value that doesn't go anywhere except ST's own backend.

LeZeez replied:
The structure itself for google isn't the same as other params, it wasn't in their docs so I just tried to figure it out until turns out this is the only way (Made 2.5 pro convert it to JS as I found it out in Python) so yeah it's not a single number.

LeZeez replied:
> You don't have to put thinkingConfig into generationConfig as an object, it's an internal value that doesn't go anywhere except ST's own backend.

I don't really think I understand, but as you said `make sure the code is not too ugly` in the Contribution guidelines, and of course the things that are NOT necessary at all will be removed + the `stop` one will be fixed. at least it works.🥲

Cohee1207 replied:
`generationConfig['thinking_budget'] = 0`

That's all you have to do. Then use this on a backend. Done.

cloak1505 commented:
Functionally, it works great.

I believe `google_master_enable_thinking` and `google_set_thinking_budget` doesn't need to be named google-specific, instead `enable_thinking` and `set_thinking_budget` to be reusable in the future if other APIs do something similar.

Looking back, Cohee asked about `show_thoughts`. This one was invented when DeepSeek was introduced and was meant to *show* thoughts rather than enable it (it still does it when disabled). Looking at 3.7 Sonnet, enabling `show_thoughts` actually enables thinking rather than making-thoughts-not-hidden, so 3.7 Sonnet could be using the "Enable thinking" `enable_thinking` option instead, ~~and rename "Request model reasoning" to "Show thoughts" to reflect its `show_thoughts` setting name.~~

Would having both ~~"Show thoughts"~~ and "Enable thinking" be clunky (talking about Sonnet, not Gemini)? I don't know how many users want the reasoning content to not be retrieved.

Edit: Never mind on the rename, that might imply a way to retrieve reasoning but hide it. "Retrieve model reasoning"?

LeZeez commented:
Thank you all for the reviews. So overall it's better, but still need some cleaning. Right? Please tell me what else does it need beside index.html and the `stop`.
@Cohee1207 @cloak1505

LeZeez committed:
Move thinking logic to backend. Fixed 'stop' and others. Some index.html edits.

cloak1505 commented:
\*Cough.\* I thought about Reasoning Effort %-based dropdown vs budget slider. I admit I don't really *need* an absolute slider (technically I didn't ask for it but it made sense to me), and "make it look like AI Studio" is a bit of an excuse. % automatically keeps the budget below max response so the user doesn't have to adjust two things.

[From comment.](https://github.com/SillyTavern/SillyTavern/pull/3880#discussion_r2051843039)

```
[ ] Reasoning Effort (tooltip: blah blah but maximum 24576 tokens.)  [High ▼]
If unchecked, the budget is automatically determined by the API.
```

This checkbox is "the second checkbox" (set thinking budget), its own setting, but seamlessly integrated with Reasoning Effort.

LeZeez commented:
> *Cough.* I thought about Reasoning Effort %-based dropdown vs budget slider. I admit I don't really _need_ an absolute slider (technically I didn't ask for it but it made sense to me), and "make it look like AI Studio" is a bit of an excuse. % automatically keeps the budget below max response so the user doesn't have to adjust two things.
> 
> [From comment.](https://github.com/SillyTavern/SillyTavern/pull/3880#discussion_r2051843039)
> 
> ```
> [ ] Reasoning Effort (tooltip: blah blah but maximum 24576 tokens.)  [High ▼]
> If unchecked, the budget is automatically determined by the API.
> ```
> 
> This checkbox is "the second checkbox" (set thinking budget), its own setting, but seamlessly integrated with Reasoning Effort.

It's not exactly an excuse, but yeah similar to AI studio is actually better. Imagine someone coming from AI studio, using ST first time then see this while they are used to the slider one. And again we did talk about this and referenced to [this issue](https://github.com/SillyTavern/SillyTavern/issues/3875). So, again, similar to AI studio is indeed good.

Using the reasoning effort may actually limit it as Google may update it to make it better with the slider and more accurate.

cloak1505 commented:
> And again we did talk about this and referenced to [this issue](https://github.com/SillyTavern/SillyTavern/issues/3875). So, again, similar to AI studio is indeed good.

There's no discussion there, and I don't see other strong opinions on why we should definitely use a slider instead of set percentages. The request was mainly to get support started.

Anthropic's console and API also takes a numerical value for budget tokens yet we're Reasoning Effort'ing it. OpenRouter lets you specify either budget or effort, default 80% of max response. (No-one asked to implement OR yet, idk when it became ready. I'll open an issue eventually along with asking for High to be 80%. Edit: #3890)

> Imagine someone coming from AI studio, using ST first time then see this while they are used to the slider one.

They may go "huh" once then figure out Effort assigns a % of max response to budget, either by reading the tooltip or noticing the budget in the terminal.

Low, 50%, and high of max response makes enough sense. E.g. if you want 2k budget, then 50% of 4k max response.

LeZeez commented:
To be honest I still believe the slider is better. And similar to AI studio is *not* entirely and an excuse. Doing this will take more time from me (as I literally suck at coding). My intent was an experiment, and here I am leaving my studies to finish this PR. 

>I'm really sorry I won't be able to continue this PR mostly, I hope you understand as I'm real busy these days and this PR was just an attempt to do something. If it's not good then someone else that has better experience than me can do it🥲

I hope you all understand. Last commit fixed most of @Cohee1207's reviews and I believe it's good enough now.

Cohee1207 commented:
Not fixed. There are still two new checkboxes, one of which is completely pointless, and they still have "google" in their naming.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -336,6 +336,11 @@ async function sendMakerSuiteRequest(request, response) {
         return response.status(400).send({ error: true });
     }
 
+    const googleThinkingBudgetModels = [
```
Why another models list?

LeZeez commented:
> Not fixed. There are still two new checkboxes, one of which is completely pointless, and they still have "google" in their naming.

I believe it's better that I close this PR and never touch ST again unless I'm better at coding at this point🥲 Sorry y'all.

I'll leave the repo until someone merges an actual better version to ST.

lyncmdos commented:
> > Not fixed. There are still two new checkboxes, one of which is completely pointless, and they still have "google" in their naming.
> 
> I believe it's better that I close this PR and never touch ST again unless I'm better at coding at this point🥲 Sorry y'all.
> 
> I'll leave the repo until someone merges an actual better version to ST.
 
it seems like i will not be able to update my ST for a long time.🥲
